src/deterministic_randomizer.rs

```diff
@@ -0,0 +1,188 @@
+/// bounded pseudorandom numbers or sampling or shuffling arrays deterministically. You can always reseed the randomizer
+/// with a different seed, or reset it to the initial seed.
+///
+/// ```edition2018
+/// # use tari_crypto::deterministic_randomizer::DeterministicRandomizer;
+/// use rand_chacha::ChaCha12Rng;
+///
+/// // In consensus-type applications, the seed might come from a hash output whose input is some kind of state
+/// // We'll just make one arbitrarily; it must be `[u8; N]` with `N <= 32`
+/// let seed = [1u8; 32];
+/// let mut randomizer = DeterministicRandomizer::<ChaCha12Rng>::new(seed);
+///
+/// // Generate bounded numbers in the range `[0, 1000)`
+/// // Because it's deterministic, we know exactly what the results should be!
+/// assert_eq!(randomizer.next_bounded_u64(1000).unwrap(), 573);
+/// assert_eq!(randomizer.next_bounded_u64(1000).unwrap(), 786);
+///
+/// // Reset the generator, which uses the original seed we provided
+/// randomizer.reset();
+/// assert_eq!(randomizer.next_bounded_u64(1000).unwrap(), 573);
+/// assert_eq!(randomizer.next_bounded_u64(1000).unwrap(), 786);
+///
+/// // We can also reseed using a new seed
+/// // This has the same effect as creating a fresh `DeterministicRandomizer` with the new seed
+/// let new_seed = [2u8; 32];
+/// randomizer.reseed(new_seed);
+///
+/// // Shuffle an array in place (which may be of any type or length)
+/// // Because it's still deterministic, we know exactly what the result should be!
+/// let mut data = [0u32, 1u32, 2u32, 3u32, 4u32];
+/// randomizer.shuffle(&mut data).unwrap();
+/// assert_eq!(&data, &[2u32, 0u32, 4u32, 3u32, 1u32]);
+///
+/// // Get a shuffled sample of an array, leaving the original unchanged
+/// let data = [0u32, 1u32, 2u32, 3u32, 4u32];
+/// let sample = randomizer.sample(&data, 3).unwrap();
+/// assert_eq!(&sample, &[3u32, 0u32, 4u32]);
+/// ```
+///
+/// **WARNING**: While `DeterministicRandomizer` requires a cryptographically-secure pseudorandom number generator
+/// (CSPRNG), it is _not_ suitable for non-deterministic use cases like key or nonce generation. If you aren't
+/// absolutely sure that you need the functionality provided here, you should use a high-entropy non-deterministic
+/// generator instead.
+#[derive(Zeroize)]
+pub struct DeterministicRandomizer<R>
+where R: SeedableRng
+{
+    prng: R,
+    seed: <R as SeedableRng>::Seed,
+}
+
+impl<R> DeterministicRandomizer<R>
+where
+    R: CryptoRng + RngCore + SeedableRng,
+    <R as SeedableRng>::Seed: Clone,
+{
+    /// Initialize the randomizer with a seed
+    pub fn new(seed: <R as SeedableRng>::Seed) -> Self {
+        Self {
+            prng: R::from_seed(seed.clone()),
+            seed,
+        }
+    }
+
+    /// Reseed the randomizer with a new seed
+    pub fn reseed(&mut self, seed: <R as SeedableRng>::Seed) {
+        self.prng = R::from_seed(seed.clone());
+        self.seed = seed;
+    }
+
+    /// Reset the randomizer to its original seed
```

Comment:
```suggestion
    /// Reset the randomizer with its current seed.
```

Since `r.seed(n); r.reseed(m); r.reset()` will use m and not n

Reply:
Done!